### ISSUE TITLE
Rust quickstart appears incomplete

### ISSUE THREAD

**Issue by michaelmior** [VivekPanyam/carton#154]
When trying the quickstart with Rust, I got several compilation errors:

```
error[E0433]: failed to resolve: use of undeclared crate or module `ndarray`
  --> src/main.rs:14:15
   |
14 |     let arr = ndarray::ArrayD::from_shape_vec(
   |               ^^^^^^^ use of undeclared crate or module `ndarray`

error[E0433]: failed to resolve: use of undeclared type `Tensor`
  --> src/main.rs:22:37
   |
22 |         .infer([("input_sequences", Tensor::<GenericStorage>::String(arr))])
   |                                     ^^^^^^ use of undeclared type `Tensor`
   |
help: consider importing this enum
   |
1  + use carton::types::Tensor;
   |

error[E0412]: cannot find type `GenericStorage` in this scope
  --> src/main.rs:22:46
   |
22 |         .infer([("input_sequences", Tensor::<GenericStorage>::String(arr))])
   |                                              ^^^^^^^^^^^^^^ not found in this scope
   |
help: consider importing this struct
   |
1  + use carton::types::GenericStorage;
   |

error[E0433]: failed to resolve: use of undeclared crate or module `ndarray`
  --> src/main.rs:15:9
   |
15 |         ndarray::IxDyn(&[1]),
   |         ^^^^^^^ use of undeclared crate or module `ndarray`

error[E0752]: `main` function is not allowed to be `async`
 --> src/main.rs:4:1
  |
4 | async fn main() {
  | ^^^^^^^^^^^^^^^ `main` function is not allowed to be `async`

Some errors have detailed explanations: E0412, E0433, E0752.
For more information about an error, try `rustc --explain E0412`.
error: could not compile `carton-test2` (bin "carton-test2") due to 5 previous errors
```

To resolve, I had to add a couple crates:

```bash
cargo add -F macros,rt-multi-thread tokio
cargo add ndarray
```

I also needed to make a few small code changes:

```diff
@@ -1,6 +1,9 @@
 use carton::Carton;
+use carton::types::GenericStorage;
 use carton::types::LoadOpts;
+use carton::types::Tensor;

+#[tokio::main]
 async fn main() {
     // Load the model
     let model = Carton::load(
```

**Comment by michaelmior** [comment]
I should note that although this was able to compile and start, it ran for a while and then produced the following error:

```
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', source/carton-runner-rust-bert/src/masked_language.rs:99:49
stack backtrace:
   0:     0x563d9870bc66 - std::backtrace_rs::backtrace::libunwind::trace::h66dc1c6acf794faa
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x563d9870bc66 - std::backtrace_rs::backtrace::trace_unsynchronized::ha80d20099a67f790
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x563d9870bc66 - std::sys_common::backtrace::_print_fmt::h7b959d43f35f16d4
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/sys_common/backtrace.rs:65:5
   3:     0x563d9870bc66 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::hdaa196410d9ee0b9
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x563d982ea52f - core::fmt::rt::Argument::fmt::h0ddfbbe8be3f80d0
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/core/src/fmt/rt.rs:138:9
   5:     0x563d982ea52f - core::fmt::write::h66b3c629f3d623e4
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/core/src/fmt/mod.rs:1094:21
   6:     0x563d98709337 - std::io::Write::write_fmt::hb6d80fba4115e0c2
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/io/mod.rs:1714:15
   7:     0x563d9870ba75 - std::sys_common::backtrace::_print::h1a49cfb0cf3cce17
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/sys_common/backtrace.rs:47:5
   8:     0x563d9870ba75 - std::sys_common::backtrace::print::hca95c2d0055e42a2
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/sys_common/backtrace.rs:34:9
   9:     0x563d9870d073 - std::panicking::default_hook::{{closure}}::hc03c01c56bca600c
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/panicking.rs:269:22
  10:     0x563d9870ce34 - std::panicking::default_hook::hb2cb5315b6634f1c
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/panicking.rs:288:9
  11:     0x563d9870d566 - std::panicking::rust_panic_with_hook::h75cd912a39a34e8a
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/panicking.rs:705:13
  12:     0x563d9870d444 - std::panicking::begin_panic_handler::{{closure}}::h1498b46f7849e167
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/panicking.rs:595:13
  13:     0x563d9870c066 - std::sys_common::backtrace::__rust_end_short_backtrace::hd36a39b27b98086b
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/sys_common/backtrace.rs:151:18
  14:     0x563d9870d202 - rust_begin_unwind
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/panicking.rs:593:5
  15:     0x563d97f12dd3 - core::panicking::panic_fmt::h98ef273141454c23
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/core/src/panicking.rs:67:14
  16:     0x563d97f12e63 - core::panicking::panic::hf53fd8b0bfa5848e
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/core/src/panicking.rs:117:5
  17:     0x563d9810b3d8 - core::option::Option<T>::unwrap::h12c43accf485d83d
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/core/src/option.rs:935:21
  18:     0x563d9810b3d8 - <carton_runner_rust_bert::masked_language::CartonMaskedLanguageModel as carton_runner_rust_bert::Model>::infer::h911637b12d75fc53
                               at /app/source/carton-runner-rust-bert/src/masked_language.rs:99:49
  19:     0x563d98235737 - carton_runner_rust_bert::main::{{closure}}::{{closure}}::haff9764bc00c2e36
                               at /app/source/carton-runner-rust-bert/src/main.rs:103:53
  20:     0x563d98235737 - core::option::Option<T>::map::h46bc3bc572d9135b
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/core/src/option.rs:1075:29
  21:     0x563d98235737 - carton_runner_rust_bert::main::{{closure}}::h22763948a8181bc0
                               at /app/source/carton-runner-rust-bert/src/main.rs:103:30
  22:     0x563d982315c1 - tokio::runtime::park::CachedParkThread::block_on::{{closure}}::hefce4379bf2999ac
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.22.0/src/runtime/park.rs:272:63
  23:     0x563d982315c1 - tokio::runtime::coop::with_budget::h948ec6343d66ed06
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.22.0/src/runtime/coop.rs:102:5
  24:     0x563d982315c1 - tokio::runtime::coop::budget::hb559616204cdf787
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.22.0/src/runtime/coop.rs:68:5
  25:     0x563d982315c1 - tokio::runtime::park::CachedParkThread::block_on::h569f593b7d92f31c
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.22.0/src/runtime/park.rs:272:31
  26:     0x563d98260bc4 - tokio::runtime::context::BlockingRegionGuard::block_on::h4f70676a06ac2c7b
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.22.0/src/runtime/context.rs:255:13
  27:     0x563d98260bc4 - tokio::runtime::scheduler::multi_thread::MultiThread::block_on::hb1b4399a43fda72d
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.22.0/src/runtime/scheduler/multi_thread/mod.rs:66:9
  28:     0x563d98260bc4 - tokio::runtime::runtime::Runtime::block_on::h9b65277f6e61fe63
                               at /root/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tokio-1.22.0/src/runtime/runtime.rs:281:45
  29:     0x563d981b89d3 - carton_runner_rust_bert::main::h9e59ea0bfb57c7cf
                               at /app/source/carton-runner-rust-bert/src/main.rs:30:5
  30:     0x563d981ad183 - core::ops::function::FnOnce::call_once::h7a80138e58956fe0
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/core/src/ops/function.rs:250:5
  31:     0x563d981ad183 - std::sys_common::backtrace::__rust_begin_short_backtrace::h8cf37ada6016b8f2
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/sys_common/backtrace.rs:135:18
  32:     0x563d981ad7c9 - std::rt::lang_start::{{closure}}::h2df01761492e40ba
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/rt.rs:166:18
  33:     0x563d98703555 - core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::h87e9a912d8a0de33
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/core/src/ops/function.rs:284:13
  34:     0x563d98703555 - std::panicking::try::do_call::ha8d57d42181c12cf
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/panicking.rs:500:40
  35:     0x563d98703555 - std::panicking::try::h2fdb2d19c253437c
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/panicking.rs:464:19
  36:     0x563d98703555 - std::panic::catch_unwind::h9fe5959d2e133449
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/panic.rs:142:14
  37:     0x563d98703555 - std::rt::lang_start_internal::{{closure}}::ha072eb7bbeaa0b6a
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/rt.rs:148:48
  38:     0x563d98703555 - std::panicking::try::do_call::h983dae29f79ed5db
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/panicking.rs:500:40
  39:     0x563d98703555 - std::panicking::try::h15014751f4b412ba
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/panicking.rs:464:19
  40:     0x563d98703555 - std::panic::catch_unwind::hf837426183b1055f
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/panic.rs:142:14
  41:     0x563d98703555 - std::rt::lang_start_internal::h2bbe0b58b2b89a9f
                               at /rustc/d5c2e9c342b358556da91d61ed4133f6f50fc0c3/library/std/src/rt.rs:148:20
  42:     0x563d981b8aac - main
  43:     0x7fc5003e3083 - __libc_start_main
                               at /build/glibc-SzIz7B/glibc-2.31/csu/../csu/libc-start.c:308:16
  44:     0x563d97f52f0e - _start
  45:                0x0 - <unknown>
```

**Comment by VivekPanyam** [comment]
Thanks for trying it out!

Convenient timing as I just put up #155 which should update the quickstart to fix these issues. I'll update it to also describe the additional required dependencies.

If you change `"input_sequences"` in your code to `"input"` it shouldn't panic any more.

Generally, incorrect tensor names shouldn't cause a panic, but we'll clean that up as the library becomes more stable. 

Happy to accept a PR if you'd like to contribute! The issue is here:

https://github.com/VivekPanyam/carton/blob/d92e847334e14446c8dee034a05ced57bb338d54/source/carton-runner-rust-bert/src/masked_language.rs#L98-L99

**Comment by VivekPanyam** [comment]
Just updated the PR to also include the additional required dependencies. Can you confirm this build of the quickstart guide looks good to you?

https://f0469c12.carton.pages.dev/quickstart

Thank you for the detailed feedback and fix! I appreciate it :)

**Comment by michaelmior** [comment]
@VivekPanyam This looks mostly alright aside from the missing dependency on `env_logger`. However, the program still panics when it runs even though it seems to get the correct output.

```
{"scores": Float([[14.551313]], shape=[1, 1], strides=[1, 1], layout=CFcf (0xf), dynamic ndim=2), "tokens": String([["day"]], shape=[1, 1], strides=[1, 1], layout=CFcf (0xf), dynamic ndim=2)}
thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 32, kind: BrokenPipe, message: "Broken pipe" }', /app/source/carton-runner-interface/src/do_not_modify/framed.rs:71:38
stack backtrace:
```

**Comment by VivekPanyam** [comment]
I edited the earlier comment; the most recent link should have the `env_logger` dependency too.

The issue for that panic is #148. The panic is actually in the runner process while it's shutting down, not in your main program. So it doesn't practically impact things, but it is confusing and a bad experience.

I'm currently working on a fix for #148 and should have it up in a few hours

**Comment by michaelmior** [comment]
Yeah, I get that. Just thought I would mention the issue. I think you can close this whenever you're ready. Thanks!

**Comment by VivekPanyam** [comment]
I appreciate it :)

I just put up a PR that fixes #148. New runners that don't panic during shutdown should be available after the nightly build runs. I'll post more details on how to get them in #148 once the PR lands

**Comment by VivekPanyam** [comment]
@michaelmior  The nightly builds are complete and new runners have been released that solve the problem. See https://github.com/VivekPanyam/carton/issues/148#issuecomment-1739930075 for instructions on updating.
